From 4c898b6a9d7296bee0df4666676152c67ec7dfe9 Mon Sep 17 00:00:00 2001
From: Marko Lindqvist <cazfi74@gmail.com>
Date: Mon, 19 Oct 2020 10:34:08 +0300
Subject: [PATCH 25/25] Don't hide allied stealth units on seen tiles

They are expected to be visible. Hiding them lead to duplicate
unit removal from client when such a unit dies

Reported by Jacob Nevins

See hrm Bug #764976

Signed-off-by: Marko Lindqvist <cazfi74@gmail.com>
---
 common/player.c  | 5 ++---
 server/maphand.c | 8 ++++++--
 2 files changed, 8 insertions(+), 5 deletions(-)

diff --git a/common/player.c b/common/player.c
index d779c1ecf4..d181e6d016 100644
--- a/common/player.c
+++ b/common/player.c
@@ -1028,7 +1028,8 @@ bool can_player_see_unit_at(const struct player *pplayer,
     } extra_type_list_iterate_end;
   }
 
-  /* Allied or non-hiding units are always seen. */
+  /* Allied or non-hiding units are always seen.
+   * See also stealth unit hiding part in map_change_seen() */
   if (pplayers_allied(unit_owner(punit), pplayer)
       || !is_hiding_unit(punit)) {
     return TRUE;
@@ -1037,8 +1038,6 @@ bool can_player_see_unit_at(const struct player *pplayer,
   /* Hiding units are only seen by the V_INVIS fog layer. */
   return fc_funcs->player_tile_vision_get(ptile, pplayer,
                                           unit_type_get(punit)->vlayer);
-
-  return FALSE;
 }
 
 /*******************************************************************//**
diff --git a/server/maphand.c b/server/maphand.c
index 78bbc2fa4e..631300c1dc 100644
--- a/server/maphand.c
+++ b/server/maphand.c
@@ -601,7 +601,7 @@ void send_tile_info(struct conn_list *dest, struct tile *ptile,
   Assumption: Each unit type is visible on only one layer.
 **************************************************************************/
 static bool unit_is_visible_on_layer(const struct unit *punit,
-				     enum vision_layer vlayer)
+                                     enum vision_layer vlayer)
 {
   return XOR(vlayer == V_MAIN, is_hiding_unit(punit));
 }
@@ -916,7 +916,11 @@ void map_change_seen(struct player *pplayer,
 
     unit_list_iterate(ptile->units, punit) {
       if (unit_is_visible_on_layer(punit, V_INVIS)
-          && can_player_see_unit(pplayer, punit)) {
+          && can_player_see_unit(pplayer, punit)
+          && (plrtile->seen_count[V_MAIN] + change[V_MAIN] <= 0
+              || !pplayers_allied(pplayer, unit_owner(punit)))) {
+        /* Allied units on seen tiles (V_MAIN) are always seen.
+         * That's how can_player_see_unit_at() works. */
         unit_goes_out_of_sight(pplayer, punit);
       }
     } unit_list_iterate_end;
-- 
2.28.0

